ui: highlight sidebar links for nested pages

couchand opened the pull request:
The route reorganization broke one of react-router's assumptions: that
logically nested pages are always on strictly-nested routes.  This
fixes the sidebar link highlighting for nested pages whose route
doesn't start with the exact route the link is going to.

Fixes: #24069
Release note: None

<img width="344" alt="screen shot 2018-04-11 at 3 46 24 pm" src="https://user-images.githubusercontent.com/793969/38639723-1a4ef4e4-3da0-11e8-98c9-d83ac070d7b0.png">

couchand committed:
ui: highlight sidebar links for nested pages

The route reorganization broke one of react-router's assumptions: that
logically nested pages are always on strictly-nested routes.  This
fixes the sidebar link highlighting for nested pages whose route
doesn't start with the exact route the link is going to.

Fixes: #24069
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/24705)
<!-- Reviewable:end -->

couchand commented:
@vilterp one question I had here was if we should just use this facility for all highlighting, rather than splitting the logic between react-router for the default behavior and this custom logic for the override.  It might be a bit simpler/more obvious to only use this and never rely on react-router at all.  What do you think?

vilterp commented on pkg/ui/src/views/app/components/layoutSidebar/index.tsx:
```diff
@@ -49,10 +71,10 @@ export default class extends React.Component<{}, {}> {
     return <nav className="navigation-bar">
       <ul className="navigation-bar__list">
-        <IconLink to="/overview" icon={homeIcon} title="Overview" />
+        <IconLink to="/overview" icon={homeIcon} title="Overview" indexFor={["/nodes", "/node"]} />
```
A comment on `IconLinkProps` explaining the semantics of `indexFor` would be nice — I assume that it does a prefix match, so e.g. `/nodes/5` will cause the `/overview` link to be highlighted? Unclear that that would work from [this SO answer about `router.isActive`](https://stackoverflow.com/questions/43750056/checking-router-isactive-on-nested-routes-in-react?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa)

couchand replied:
yes, I spent a little while trying to figure out an obvious name for the prop... just changed it to `activeFor`, do you think that's clear enough, particularly after unifying all active class logic?

vilterp reviewed (approved):
Interesting question about using react-router's `activeClassName` thing, which relies on the nested-routes assumption we're breaking, vs. this new thing you came up with.

I guess my opinion is that I wish we weren't breaking that assumption — that there actually was a `nodes/` index and a `nodes/:id`. But that ship has sailed — the overview is the node list now (as well as showing other non-node things like ranges), and we've removed the standalone node list. Maybe some way of solving this will occur to us as part of our IA discussion.

Until we think of an elegant solution, I think what you have now is fine. I'd just request that you add a little comment explaining why `indexFor` is necessary (something like what you have in the PR description) and how its semantics (left a comment about the specific part that confused me).

couchand committed:
ui: cleanup IconLink props

Remove unused onlyActiveOnIndex and require icon.

Release note: None

couchand committed:
ui: only use custom logic for sidebar link active class

This simplifies the IconLink active class logic by unifying the
cases to always use the custom, more obvious form.

Release note: None

couchand commented on pkg/ui/src/views/app/components/layoutSidebar/index.tsx:
```diff
@@ -1,4 +1,7 @@
+import classnames from "classnames";
```
done

vilterp reviewed (approved):
LGTM.

Don't have a strong opinion here, but why `/node/<n>` instead of `/nodes/<n>`?

couchand commented on pkg/ui/src/views/app/components/layoutSidebar/index.tsx:
```diff
@@ -23,14 +26,25 @@ interface IconLinkProps {
     className: "normal",
-    onlyActiveOnIndex: false,
+    activeFor: [] as string[],
+  };
```
good point

couchand replied:
Partial<IconLinkProps> to optionalize everything but otherwise great!

couchand commented:
bors r+

couchand commented:
> why `/node/<n>` instead of `/nodes/<n>`?

Mainly to reserve the collection namespace for interesting sub-collections.  For instance, `/nodes/decommissioned` or some such.

vilterp commented:
Interesting, hadn't thought of the reserving the namespace consideration.
